src/IssueViz.Security.UnitTests/Hotspots/LocalHotspotStoreTests.cs

```diff
@@ -0,0 +1,343 @@
+            .BeEquivalentTo(new IssuesChangedEventArgs(Array.Empty<IAnalysisIssueVisualization>(), hotspots));
+    }
+
+    [TestMethod]
+    public void AddOrUpdate_NoServerHotspots_UpdatesForSameFile()
+    {
+        var issueVis1 = Mock.Of<IAnalysisIssueVisualization>();
+        var issueVis2 = Mock.Of<IAnalysisIssueVisualization>();
+        var issueVis3 = Mock.Of<IAnalysisIssueVisualization>();
+        var testSubject = CreateTestSubject(out var eventListener);
+        var oldHotspots = new[] { issueVis1 };
+        testSubject.AddOrUpdate("lalala", oldHotspots);
+        var newHotspots = new[] { issueVis2, issueVis3 };
+
+        testSubject.AddOrUpdate("lalala", newHotspots);
+
+        VerifyContent(testSubject, newHotspots.Select(x => new LocalHotspot(x)).ToArray());
+        eventListener.Events.Should().HaveCount(2).And.Subject.Last().Should()
+            .BeEquivalentTo(new IssuesChangedEventArgs(oldHotspots, newHotspots));
+    }
+
+    [TestMethod]
+    public void AddOrUpdate_NoServerHotspots_AddsForDifferentFile()
+    {
+        var issueVis1 = Mock.Of<IAnalysisIssueVisualization>();
+        var issueVis2 = Mock.Of<IAnalysisIssueVisualization>();
+        var testSubject = CreateTestSubject(out var eventListener);
+        testSubject.AddOrUpdate("lalala", new[] { issueVis1 });
+        var newHotspots = new[] { issueVis2 };
+
+        testSubject.AddOrUpdate("cococo", newHotspots);
+
+        VerifyContent(testSubject, new LocalHotspot(issueVis1), new LocalHotspot(issueVis2));
+        eventListener.Events.Should().HaveCount(2).And.Subject.Last().Should()
+            .BeEquivalentTo(new IssuesChangedEventArgs(Array.Empty<IAnalysisIssueVisualization>(), newHotspots));
+    }
+
+    [TestMethod]
+    public void AddOrUpdate_ServerHotspots_MatchesCorrectly()
+    {
+        var serverStoreMock = new Mock<IServerHotspotStore>();
+        var serverHotspot1 = CreateEmptyServerHotspot();
+        var serverHotspot2 = CreateEmptyServerHotspot();
+        serverStoreMock.Setup(x => x.GetAll()).Returns(new[] { serverHotspot1, serverHotspot2 });
+
+        var issueVis3 = Mock.Of<IAnalysisIssueVisualization>();
+        var issueVis1 = Mock.Of<IAnalysisIssueVisualization>();
+        var issueVis2 = Mock.Of<IAnalysisIssueVisualization>();
+
+        var matcherMock = new Mock<IHotspotMatcher>();
+        matcherMock.Setup(x => x.IsMatch(issueVis1, serverHotspot1)).Returns(true);
+        matcherMock.Setup(x => x.IsMatch(issueVis2, serverHotspot2)).Returns(true);
+        matcherMock.Setup(x => x.IsMatch(issueVis3, serverHotspot1))
+            .Returns(true); // 2 local hotspots match 1 server but only one pair should be created
+
+        var testSubject = CreateTestSubject(out _, serverStoreMock.Object, matcherMock.Object);
+
+        testSubject.AddOrUpdate("lalala", new[] { issueVis1, issueVis2, issueVis3 });
+
+        VerifyContent(testSubject, new LocalHotspot(issueVis1, serverHotspot1),
```

Comment:
Minor: listing the expected values on separate lines would make them easier to read and compare to the code above.